etcd: use Unix Domain socket

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind flake

#### What this PR does / why we need it:

Choosing a port in advance is racy. A better solution is to use a Unix Domain
socket in the per-etcd-instance data directory. Then the name can be determined
in advance and there's no risk of conflicts with other etcd instances.

### Notes to reviewers
    
With unix:// for the endpoint, we have to be a bit more careful about
passing a TLS config to the etcd client library because for unix://, in
contrast to http://, it tries to use an incomplete config which
then fails to establish the connection.



#### Which issue(s) this PR fixes:
Related-to: https://kubernetes.slack.com/archives/C09QZ4DQB/p1744360811926169

```

    === RUN   TestSchedulerPerf/SteadyStateClusterResourceClaimTemplateFirstAvailable/fast
    {"level":"warn","ts":"2025-04-11T03:32:06.676527Z","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
    {"level":"warn","ts":"2025-04-11T03:32:06.676707Z","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
    {"level":"warn","ts":"2025-04-11T03:32:06.677056Z","caller":"etcdmain/etcd.go:146","msg":"failed to start etcd","error":"listen tcp 127.0.0.1:37803: bind: address already in use"}
    {"level":"fatal","ts":"2025-04-11T03:32:06.677104Z","caller":"etcdmain/etcd.go:204","msg":"discovery failed","error":"listen tcp 127.0.0.1:37803: bind: address already in use","stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\tgo.etcd.io/etcd/server/v3/etcdmain/etcd.go:204\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\tgo.etcd.io/etcd/server/v3/etcdmain/main.go:40\nmain.main\n\tgo.etcd.io/etcd/server/v3/main.go:31\nruntime.main\n\truntime/proc.go:272"}
       etcd.go:260: unable to start etcd: could not start etcd
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @BenTheElder

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -111,14 +130,12 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
 	}
-	klog.Infof("storing etcd data in: %v", etcdDataDir)
 
+	logger.V(6).Info("starting etcd", "url", customURL, "dataDir", etcdDataDir)
```
Was URL and directory really important enough to be logged at V(0)? I don't think so. V(6) is debatable, it could also be higher.

Overall I am making the code a lot less verbose.

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -136,13 +153,65 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
+	reader, writer := io.Pipe()
+	cmd.Stdout = writer
+	cmd.Stderr = writer
+	var portInUse atomic.Bool
```
Actually checking this is missing. While writing a test for it I noticed that the existing code which connects to the port doesn't detect that *someone else* took over the port -> more work needed, WIP.

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -136,62 +165,154 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
-		if i == pollCount {
+	// We need to wait until etcd is ready. Connecting to its port is not sufficient for that
+	// because something else may have grabbed the port before our etcd instance did.
+	// Instead we rely on etcd output to determine when it is ready.
```
The hack above with checking for "simple token is not cryptographically signed" is gross.

But this code here is also broken and completely fails to detect the scenario that two integration tests try to start etcd on the same port. One etcd starts, one fails, and then both integration tests use **the same** etcd?!

While testing, I got this error from etcd:
```
2025/04/11 16:30:39 WARNING: [core] [Channel #3 SubChannel #4] grpc: addrConn.createTransport failed to connect to {Addr: "127.0.0.1:34847", ServerName: "127.0.0.1:34847", }. Err: connection error: desc = "error reading server preface: EOF"
```

I am also not sure whether all etcd instances are properly ignoring each other. What does `--listen-peer-urls http://127.0.0.1:0` (note the unspecified port 0) do?

What are the chances that we can get Unix Domain socket support in etcd? Then we could drop all this fragile "try to guess port in advance" crap.

cc @serathius

pohly replied:
https://github.com/etcd-io/etcd/issues/12450

serathius replied:
Binding TCP socket to port zero will return you any available port. This is a easy way to ensure you don't have port conflicts. 

> What are the chances that we can get Unix Domain socket support in etcd? Then we could drop all this fragile "try to guess port in advance" crap.

Don't think unix sockets remove the problem of conflict, instead of conflicting on port number, you have conflict on paths.

BenTheElder replied:
> Binding TCP socket to port zero will return you any available port. This is a easy way to ensure you don't have port conflicts

As long as you use this socket all the way through. If you use it to find a port then pass it to Etcd startup seperately there is still a race.

BenTheElder replied:
Are we able to tell ETCD to use 0 and then get the selected port?

pohly replied:
> Binding TCP socket to port zero will return you any available port.

So that is what `--listen-peer-urls http://127.0.0.1:0` does, right? But who is then connecting to the dynamically assigned port? How do they get to know what it is?

> Don't think unix sockets remove the problem of conflict, instead of conflicting on port number, you have conflict on paths.

That's what a temporary directory is for: the caller can create that atomically, then no other instance is going to try and use the same socket path.

> Are we able to tell ETCD to use 0 and then get the selected port?

No, not at the moment. We could add log output for it, but it then should be of a level higher than "info", because enabling that produced additional output. I think the non-JSON gRPC output was enabled by it.

Instead of adding such a feature it would be easier to backport the [support for Unix Domain sockets](https://github.com/etcd-io/etcd/pull/12469).

pohly replied:
> As long as etcd will accept it, we can detect the bound port (via logs, or by inspecting the process).

I already considered inspecting the process, but then shied away from what be a Linux-only solution (inspect procfs). Extending logging would be portable.

But back to my my question: what is `--listen-peer-urls http://127.0.0.1:0` good for? Is this feature perhaps simply not used (because no-one knows the port) and has to be configured like this because it cannot be disabled?

> Yes, though that also means the integration tests are less realistic, since that's unlikely to be what production api-server + etcd looks like.

Both Unix Domain and TCP via localhost are going to be handled entirely inside the kernel. I suppose performance might be a bit different, but we already deviate from real production usage a lot more by a wide margin by putting the data directory into /tmp.... I'm not worried about using Unix Domain sockets for testing.

pohly replied:
> I already considered inspecting the process, but then shied away from what be a Linux-only solution (inspect procfs).

Also, `--advertise-client-urls` probably won't work (etcd would know that it needs to inject the port, which I doubt it does). I don't know whether that matters.

pohly replied:
I need to clarify: we can use `--listen-client-urls` and `--advertise-client-urls` with a pre-chosen port (like we do now) and then check via procfs that the etcd instance is listening on it. If we use `--listen-client-urls ...:0`, then `--advertise-client-urls` won't work.

pohly replied:
@aojea wrote: 
> I really prefer to explore other options than use this, seems very brittle

See discussion above.

I prefer to use Unix Domain sockets. They don't have any of these issues.

Two options:
- Backport https://github.com/etcd-io/etcd/pull/12469 to etcd 3.5 and use a release with that backport.
- Release 3.6 and use it in Kubernetes.

Until then we can either do nothing and accept that the current code flakes or use this PR as stop-gap solution.

@serathius: I have seen some activity around etcd 3.6. Is the goal to migrate Kubernetes master to it as soon as it is released? Is that going to happen for Kubernetes 1.34? If yes, then a backport doesn't make sense. Otherwise it does.

aojea replied:
is not available in 3.5 https://github.com/etcd-io/etcd/pull/15940 ?

pohly replied:
I missed that it was backported because the issue didn't mention it and wasn't referenced. Unfortunately it's only available in etcd >= [v1.3.29](https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.5.md#v3513-2024-03-29). We are on 3.5.17, so we still need a bump of etcd.

pohly replied:
@serathius: how is the registry.k8s.io/etcd:3.5.29-0 image built? The release was tagged, but there's no such image yet.

pohly replied:
I can't read! It's available since 3.5.14, 29 is the release date.

But it doesn't work. I had tried it earlier and double-checked right now:

```console
$ /nvme/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd --data-dir /tmp/integration_test_etcd_data4256280294 --listen-client-urls unix:///tmp/etcd.sock --advertise-client-urls unix:///tmp/etcd.sock --listen-peer-urls http://127.0.0.1:0 -log-level warn --quota-backend-bytes 8589934592
{"level":"warn","ts":"2025-04-13T17:10:21.777639+0200","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
{"level":"warn","ts":"2025-04-13T17:10:21.778829+0200","caller":"etcdmain/etcd.go:75","msg":"failed to verify flags","error":"--advertise-client-urls \"unix:///tmp/etcd.sock\" must be \"host:port\" (missing port in address)"}

$ /nvme/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd --data-dir /tmp/integration_test_etcd_data4256280294 --listen-client-urls unix://tmp/etcd.sock --advertise-client-urls unix://tmp/etcd.sock --listen-peer-urls http://127.0.0.1:0 -log-level warn --quota-backend-bytes 8589934592
{"level":"warn","ts":"2025-04-13T17:10:34.695781+0200","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
{"level":"warn","ts":"2025-04-13T17:10:34.695923+0200","caller":"etcdmain/etcd.go:75","msg":"failed to verify flags","error":"--advertise-client-urls \"unix://tmp/etcd.sock\" must be \"host:port\" (address tmp: missing port in address)"}

$ /nvme/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd --version
etcd Version: 3.5.17
Git SHA: 507c0de
Go Version: go1.22.9
Go OS/Arch: linux/amd64
```

pohly replied:
etcd can *listen* on a Unix Domain socket, but it cannot *advertise* it. Looks like https://github.com/etcd-io/etcd/pull/12469 was incomplete:

```console
$ /nvme/gopath/src/k8s.io/kubernetes/third_party/etcd/etcd --data-dir /tmp/integration_test_etcd_data4256280294 --listen-client-urls unix:///tmp/etcd.sock --advertise-client-urls http://127.0.0.1:32432 --listen-peer-urls http://127.0.0.1:0 -log-level warn --quota-backend-bytes 8589934592
{"level":"warn","ts":"2025-04-13T19:34:27.705959+0200","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
{"level":"warn","ts":"2025-04-13T19:34:27.706876+0200","caller":"embed/config.go:689","msg":"Running http and grpc server on single port. This is not recommended for production."}
{"level":"warn","ts":"2025-04-13T19:34:27.745845+0200","caller":"auth/store.go:1241","msg":"simple token is not cryptographically signed"}
```

But I suspect that it doesn't matter for our usage.

pohly replied:
I switched to using Unix Domain sockets. All integration tests passed.

I think this is worthwhile because ["unable to start etcd: could not start etcd"](https://storage.googleapis.com/k8s-triage/index.html?text=bind%3A%20address%20already%20in%20use) is only the tip of the iceberg: for that to happen, the chosen port must be free when the helper picks it, in use when etcd starts, but not accepting connections when the helper checks for etcd running.

A much more likely failure is that the helper in test binary A connects to the *wrong* etcd started by test binary B, which then also gets used by the tests in A. In the best case, the two clients don't interfere and we don't notice. But if B terminates while A is still using the etcd instance, then we should encounter all kinds of weird errors in A.

aojea replied:
I'm supportive since in integration we don't exercise the etcd networking bits and socket consumption and tcp buffers are usually causing performance problems IIRC, @liggitt for double check

BenTheElder replied:
sidestepping performance problems in integration tests sounds a bit dubious to me, otherwise SGTM.

pohly replied:
We are already sidestepping performance problems by putting the data into /tmp.

The way I see it, integration testing *of Kubernetes* is better served by running etcd as fast and reliable as we can make it. That way we don't get random flakes in tests that the test owners then most likely cannot trace back to an issue with etcd.

Integration testing *of etcd* must use a deployment of etcd that is as close to a real cluster as possible. In Kubernetes, E2E testing is closer to that than integration testing.

I can also see the other side of the coin: run etcd as close to a real cluster as possible to increase the chance that some bug in etcd is triggered. But then we shouldn't use /tmp.

BenTheElder replied:
Yeah, agreed. We just often wind up depending on implicit coverage between these scenarios.
I've considered having kind e2e also use etcd in tmpfs to avoid performance issues ... https://github.com/kubernetes-sigs/kind/issues/845#issuecomment-1261248420

though, this needs measuring again to check against reality, in CI depending on the hardware and workload it's not necessarily much different (eg a lot of CI is using local SSD that are rather fast)

serathius replied:
I don't I'm keeping up. When we started being concerned about performance? Can we split the PR so we can focus on single decision at a time? 

For context I have moved etcd integrations tests off unix socket, because how hard it was to understand differences between TCP and TLS.

pohly replied:
> When we started being concerned about performance?

@BenTheElder raised it as a concern in https://github.com/kubernetes/kubernetes/pull/131259#discussion_r2045202418. But I think we are all on the same page that in Kubernetes, using Unix Domain sockets is fine.

> Can we split the PR so we can focus on single decision at a time?

I could split out the logging part if that's what reviewers prefer.

> Faster is nice, but now when you sacrifice simplicity.

Faster is not the goal of this PR. Simplicity and race-free startup is.

BenTheElder replied:
I think it makes sense to split logging and the socket change, it's easier to reason about the socket change and that part should deflake. The logging change looks more likely to lead to bugs and likely to need more careful code review, even though I haven't spotted any yet :-)

The socket change is pretty simple code-wise, and I think we're generally agreed it's reasonable. I think we just need to be mindful that other suites must cover "realistic" etcd, but that was never a key purpose of these tests ...

pohly replied:
Done, only the Unix Domain socket part remains here.

aojea commented on test/integration/etcd/start/start_test.go:
```diff
@@ -0,0 +1,55 @@
+			framework.StartEtcd(tCtx.Logger().WithName(fmt.Sprintf("instance-%d", i)), t, true)
+			// Increase the risk of port collisions by also listening on several ports ourselves.
+			for e := 0; e < 10; e++ {
+				l, err := net.Listen("tcp", "127.0.0.1:0")
```
this leaves sockers lingering for 1 minute during the TIME_WAIT period before they can be reused, exhausting resources, use SO_LINGER https://github.com/kubernetes/kubernetes/blob/b15dfce6cbd0d5bbbcd6172cf7e2082f4d31055e/pkg/probe/dialer_others.go#L35-L39

BenTheElder replied:
xref: https://github.com/kubernetes-sigs/kind/blob/1e95d05eeccdf615e24d7af11276e2954fe265a2/pkg/cluster/internal/providers/common/getport.go#L4

aojea commented on test/integration/framework/etcd.go:
```diff
@@ -98,7 +101,22 @@ func init() {
 // RunCustomEtcd starts a custom etcd instance for test purposes.
-func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url string, stopFn func(), err error) {
+func RunCustomEtcd(logger klog.Logger, dataDir string, customFlags []string) (url string, stopFn func(), err error) {
+	for i := 0; i < 100; i++ {
```
we need to be very conservative opening network sockets, if something is wrong a few times then we need to fail because that is sympton something is wrong, at max use 5

pohly replied:
Obsoleted, using Unix Domain sockets now.

aojea commented:
I really prefer to explore other options than use this, seems very brittle

pohly committed:
etcd: use Unix Domain socket for testserver

Choosing a port in advance is racy. A better solution is to use a Unix Domain
socket in the per-etcd-instance data directory. Then the name can be determined
in advance and there's no risk of conflicts with other etcd instances.

With unix:// for the endpoint, we have to be a bit more careful about
passing a TLS config to the etcd client library because for unix://, in
contrast to http://, it tries to use an incomplete config which
then fails to establish the connection.

pohly commented:
/test pull-kubernetes-integration

serathius commented:
cc @liggitt

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -136,35 +138,88 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
 		err := cmd.Wait()
+		logger.V(2).Info("etcd exited", "err", err)
+		// Tell goroutine that we are done.
+		cancel()
```
Not canceling the context before waiting for the goroutine to exit caused the goroutine to always run into the timeout and the "etcd didn't exit in 5 seconds, killing it" log message.

pohly commented on test/integration/utils.go:
```diff
@@ -67,32 +66,7 @@ func WaitForPodToDisappear(podClient coreclient.PodInterface, podName string, in
-	}
-
-	return c, clientv3.NewKV(c), nil
+	return kubeapiservertesting.GetEtcdClients(config)
```
I moved the code to avoid adding k8s.io/kubernetes/test/integration as dependency to `cmd/kube-apiserver/app/testing`.

pohly commented:
/test pull-kubernetes-e2e-kind

https://github.com/kubernetes/kubernetes/issues/131222 ?

pull-kubernetes-linter-hints warns about the deprecated `grpc.WithBlock`, but as I am just moving the code I am going to keep it.

aojea commented on test/integration/framework/etcd.go:
```diff
@@ -136,35 +138,89 @@ func RunCustomEtcd(dataDir string, customFlags []string, output io.Writer) (url
-		cmd.Stderr = output
+
+	// Always filter the etcd output. This allows us to get rid of known harmless messages and (more importantly)
+	// to detect the "listen tcp 127.0.0.1:37803: bind: address already in use" error.
```
where is this error coming?

pohly replied:
This comment is obsolete. The remaining purpose of output interception is making the output nicer, not detecting certain events.

aojea commented on test/integration/framework/etcd.go:
```diff
@@ -97,36 +90,45 @@ func init() {
+		// in the unlikely case that something does use this URL after all, an invalid URL is set here.
 		"--advertise-client-urls",
-		customURL,
+		"http://127.0.0.111:0",
```
but does 127.0.0.111 exist?

pohly replied:
It doesn't matter, the whole point is that this can't be the real Unix Domain socket because etcd doesn't allow it and then better should be something which is guaranteed to be invalid (because of the 0 port number). This is not for listening, it's for clients.

pohly replied:
It has a default value of `http://localhost:2379` which then would be used.

None of this matters unless some client somehow picks up the advertised URL and connects to it. This should never happen as far as I understand it. But if it happens, I'd rather have that client fail than potentially connect to some actual etcd instance.

Not important, I can drop it.

BenTheElder replied:
Maybe we should pick an address we _know_ won't work but will technically be a valid value. Off the top of my head one of the well known DNS resolver IPs could work...

pohly replied:
> if we can't unset it to ""

I checked, `--advertise-client-urls=` didn't override the default.

> Maybe we should pick an address we know won't work but will technically be a valid value.

The sooner a client fails which tries to use the advertised URL the better, IMHO. Therefore I prefer an invalid address.

BenTheElder replied:
I think 127.0.0.111 meets "pretty unlikely" but not "all but guaranteed". Something like 8.8.8.8 might make more sense.

I also thought about an invalid domain but IIRC Etcd doesn't let you advertise a hostname, only IP?

pohly replied:
I misspoke: I meant "invalid URL", not "invalid address". The `:0` port is unusable for a connect call.

aojea commented:
I could only do a quick pass but looks awesome and promising

BenTheElder commented:
> I could only do a quick pass but looks awesome and promising

Same, though your comment about TCP performance issues sounds a little concerning, can you elaborate?

At least we also have the scale periodics.

cici37 commented:
/triage accepted

aojea commented:
> Same, though your comment about TCP performance issues sounds a little concerning, can you elaborate?

https://news.ycombinator.com/item?id=37466475

liggitt commented:
/lgtm
/approve
/milestone v1.34

pohly commented:
/skip pull-kubernetes-linter-hints

No need to retest, I am intentionally ignoring the `grpc.WithBlock` deprecation (avoiding it is a bigger conceptual change).

pohly commented:
/hold cancel
/skip

Let's see whether this stops retesting.
